### PR TITLE
Add spire healthcheck timeout

### PR DESCRIPTION
**Pull request by Bolodya1997**
Spire agent fails to start in 10 attempts while testing on packet setup:
```
...
Sep 29 01:56:12.171 [INFO] [cmd:spire-agent healthcheck] Agent is unavailable.
Sep 29 01:56:12.170 [INFO] [cmd:spire-agent run] time="2020-09-29T01:56:12Z" level=warning msg="Current umask 0022 is too permissive; setting umask 0027."
Sep 29 01:56:12.192 [INFO] [cmd:spire-agent healthcheck] Agent is unavailable.
Sep 29 01:56:12.170 [INFO] [cmd:spire-agent run] time="2020-09-29T01:56:12Z" level=warning msg="Insecure bootstrap enabled; skipping server certificate verification" subsystem_name=attestor
Sep 29 01:56:12.214 [INFO] [cmd:spire-agent healthcheck] Agent is unavailable.
Sep 29 01:56:12.230 [INFO] [cmd:spire-agent healthcheck] Agent is unavailable.
Sep 29 01:56:12.253 [INFO] [cmd:spire-agent healthcheck] Agent is unavailable.
Sep 29 01:56:12.270 [INFO] [cmd:spire-agent healthcheck] Agent is unavailable.
Sep 29 01:56:12.286 [INFO] [cmd:spire-agent healthcheck] Agent is unavailable.
Sep 29 01:56:12.312 [INFO] [cmd:spire-agent healthcheck] Agent is unavailable.
Sep 29 01:56:12.331 [INFO] [cmd:spire-agent healthcheck] Agent is unavailable.
Sep 29 01:56:12.354 [INFO] [cmd:spire-agent healthcheck] Agent is unavailable.
--- FAIL: TestForwarderTestSuite (0.40s)
    main_test.go:81: 
        	Error Trace:	main_test.go:81
        	            				suite.go:118
        	            				main_test.go:165
        	Error:      	"%!s(<-chan error=0xc00033aa20)" should have 0 item(s), but has 1
        	Test:       	TestForwarderTestSuite
FAIL
FAIL	github.com/networkservicemesh/cmd-forwarder-sriov	0.411s
```
With timeouts added spire agents starts:
```
Sep 29 02:32:10.254 [WARN] [cmd:spire-agent healthcheck] Agent is unavailable.
Sep 29 02:32:10.254 [INFO] [cmd:spire-agent run] time="2020-09-29T02:32:10Z" level=warning msg="Insecure bootstrap enabled; skipping server certificate verification" subsystem_name=attestor
Sep 29 02:32:10.381 [WARN] [cmd:spire-agent healthcheck] Agent is unavailable.
Sep 29 02:32:10.254 [INFO] [cmd:spire-agent run] time="2020-09-29T02:32:10Z" level=error msg="No identity issued" error="<nil>" registered=false subsystem_name=workload_api
Sep 29 02:32:10.505 [INFO] [cmd:spire-agent healthcheck] Agent is healthy.
```

### CODE REVIEW

**Comment by edwarnicke** [comment]
Did you ever figure out why the packet.net spire is taking so much longer to be healthy?

**Comment by Bolodya1997** [comment]
I am not sure that it is much longer...
* https://github.com/networkservicemesh/cmd-forwarder-sriov/pull/18/checks?check_run_id=1179792050: 220 ms to start
```
Sep 29 01:57:21.612 [INFO] [cmd:spire-agent healthcheck] Agent is unavailable.
Sep 29 01:57:21.686 [INFO] [cmd:spire-agent healthcheck] Agent is unavailable.
Sep 29 01:57:21.750 [INFO] [cmd:spire-agent healthcheck] Agent is unavailable.
Sep 29 01:57:21.820 [INFO] [cmd:spire-agent healthcheck] Agent is unavailable.
Sep 29 01:57:21.611 [INFO] [cmd:spire-agent run] time="2020-09-29T01:57:21Z" level=error msg="No identity issued" error="<nil>" registered=false subsystem_name=workload_api
Sep 29 01:57:21.839 [INFO] [cmd:spire-agent healthcheck] Agent is healthy.
```
* packet fails after 180 ms without timeout
* packet starts in 250 ms with timeout

File: pkg/tools/spire/start.go [hunk 1]
```diff
@@ -154,12 +163,16 @@ func Start(options ...Option) <-chan error {
...
+	cmdStr = fmt.Sprintf("spire-agent healthcheck -socketPath %s", spireSocketPath)
+	_ = try.Do(func(attempts int) (bool, error) {
+		err = exechelper.Run(cmdStr,
+			exechelper.WithStdout(log.Entry(opt.ctx).WithField("cmd", "spire-agent healthcheck").WriterLevel(logrus.InfoLevel)),
+			exechelper.WithStderr(log.Entry(opt.ctx).WithField("cmd", "spire-agent healthcheck").WriterLevel(logrus.WarnLevel)),
 		)
+		if err != nil {
+			<-time.After(healthCheckTimeout)
+		}
+		return true, err
```
Comment (edwarnicke) [inline pkg/tools/spire/start.go:175]: Could we please just up the number of attempts here?  I really don't care how large we make them... but timers really really aren't the answer.   We see similar timing characteristics on the Github Actions CI (which is slow) and the Packet boxes (which should be fast... but have roughly the same spread of timing for the healthcheck we see in CI).

 This approach in particular is going to slow this down *everywhere*.  100ms is already the same size I see for *total* startup time in cmd-forwarder-vppagent, so this PR just literally *doubled* our startup time... for no real benefit I can see that can't be achieved by upping the count (a much simpler change).

**Review by edwarnicke — CHANGES_REQUESTED** [review]

File: pkg/tools/spire/start.go (hunk 1 above)
Comment (denis-tingajkin) [reply pkg/tools/spire/start.go:175]: @edwarnicke why we can't just replace tries on `<-context.Done()`? In this case, we'll not care about the count of tries and our docker tests will able to set a custom expecting time for the startup of the spire. Is it make sense?

Comment (edwarnicke) [reply pkg/tools/spire/start.go]: @denis-tingajkin I like this idea very much :)

A simple loop with a

```go
for {
 exechelper.Run(
    fmt.Sprintf("spire-server healthcheck -registrationUDSPath %s/spire-registration.sock", spireRoot),
    exechelper.WithStdout(log.Entry(opt.ctx).WithField("cmd", "spire-server healthcheck").Writer()),
    exechelper.WithStderr(log.Entry(opt.ctx).WithField("cmd", "spire-server healthcheck").Writer()),
  )
  select {
  case <-opt.ctx.Done():
    errCh <- opt.ctx.Err()
    close(errCh)
    return errCh
  default:
}
```

would do it :)

Comment (edwarnicke) [reply pkg/tools/spire/start.go]: @denis-tingajkin Of course this means I now have to set a responsible timeout in my tests ;)

**Review by denis-tingajkin — COMMENTED** [review]
Only one moment I see here.

We should properly handle a case when the context has not a deadline.

**Review by edwarnicke — APPROVED** [review]

**Pull request merged**